contentpal/src/main/java/org/dmfs/android/contentpal/operations/Assert.java

```diff
@@ -0,0 +1,89 @@
+/*
+ * Copyright 2017 dmfs GmbH
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.dmfs.android.contentpal.operations;
+
+import android.content.ContentProviderOperation;
+import android.support.annotation.NonNull;
+
+import org.dmfs.android.contentpal.Operation;
+import org.dmfs.android.contentpal.Predicate;
+import org.dmfs.android.contentpal.RowData;
+import org.dmfs.android.contentpal.SoftRowReference;
+import org.dmfs.android.contentpal.Table;
+import org.dmfs.android.contentpal.TransactionContext;
+import org.dmfs.android.contentpal.predicates.AnyOf;
+import org.dmfs.android.contentpal.rowdata.EmptyRowData;
+import org.dmfs.android.contentpal.tools.uriparams.EmptyUriParams;
+import org.dmfs.optional.Absent;
+import org.dmfs.optional.Optional;
+
+
+/**
+ * An {@link Operation} for an assert query which checks that the given {@link RowData} is present in the given {@link Table}
+ * filtered by the given {@link Predicate}.
+ *
+ * @author Gabor Keszthelyi
+ */
+public final class Assert<T> implements Operation<T>
```

Comment:
We need to distinguish two different assert operations.

1. an `Assert` which allows you to specify any `Table` and `Predicate` (like `BulkUpdate` and `BulkDelete`)
2. an `Assert` which takes a `RowSnapshot` and operates on this row only (just like with `Put` and `Delete` operations).

The latter one is helpful when you want to do this "compare & swap" thing with a specific `RowSnapshot` like this
```java
new Joined
(
   new Assert(rowToUpdate, new VersionRowData(version)),
   new Put(rowToUpdate, rowData)
)
```
This `Put` operation would fail if the version number column of that row has any value other than `version`.
Operations of this type are actually used by the contacts app when you save the contact.

Reply:
So I guess we have the 1. version with the current `Assert` class, right?
For 2., we need to extend `RowReference` with `assertOperationBuilder()` method?

Can you please suggest names for these 2 Assert classes?

Reply:
We can keep this one as it is. The second one could be called `AssertRow`.
`assertOperationBuilder()` as discussed on Slack.

Reply:
Ok, thanks, it's pushed.

Reply:
Hmm ... on a second though it's probably better to keep the names in line with the other operations. So `Assert` -> `BulkAssert` and `AssertRow` -> `Assert`. What do you think?

Reply:
Yes, after pointing out the similarities with the existing one, I think, too, that it's better that way. I've renamed them.